chainbus/bus_test.go

```diff
@@ -0,0 +1,175 @@
+
+import (
+	"sync/atomic"
+	"testing"
+	"time"
+
+	bi "github.com/CovenantSQL/CovenantSQL/blockproducer/interfaces"
+)
+
+func TestNew(t *testing.T) {
+	bus := New()
+	if bus == nil {
+		t.Log("New EventBus not created!")
+		t.Fail()
+	}
+}
+
+func TestHasCallback(t *testing.T) {
+	bus := New()
+	bus.Subscribe(bi.TransactionType(1), func() {})
+	if bus.HasCallback(bi.TransactionType(2)) {
+		t.Fail()
+	}
+	if !bus.HasCallback(bi.TransactionType(1)) {
+		t.Fail()
+	}
+}
+
+func TestSubscribe(t *testing.T) {
+	bus := New()
+	if bus.Subscribe(bi.TransactionType(1), func() {}) != nil {
+		t.Fail()
+	}
+	if bus.Subscribe(bi.TransactionType(1), "String") == nil {
+		t.Fail()
+	}
+}
+
+func TestSubscribeOnce(t *testing.T) {
+	bus := New()
+	if bus.SubscribeOnce(bi.TransactionType(1), func() {}) != nil {
+		t.Fail()
+	}
+	if bus.SubscribeOnce(bi.TransactionType(1), "String") == nil {
+		t.Fail()
+	}
+}
+
+func TestSubscribeOnceAndManySubscribe(t *testing.T) {
+	bus := New()
+	event := bi.TransactionType(1)
+	flag := 0
+	fn := func() { flag += 1 }
+	bus.SubscribeOnce(event, fn)
+	bus.Subscribe(event, fn)
+	bus.Subscribe(event, fn)
+	bus.Publish(event)
+
+	if flag != 3 {
+		t.Fail()
+	}
+}
+
+func TestUnsubscribe(t *testing.T) {
+	bus := New()
+	handler := func() {}
+	bus.Subscribe(bi.TransactionType(1), handler)
+	if bus.Unsubscribe(bi.TransactionType(1), handler) != nil {
+		t.Fail()
+	}
+	if bus.Unsubscribe(bi.TransactionType(1), handler) == nil {
+		t.Fail()
+	}
+}
+
+func TestPublish(t *testing.T) {
+	bus := New()
+	bus.Subscribe(bi.TransactionType(1), func(a int, b int) {
+		if a != b {
+			t.Fail()
+		}
+	})
+	bus.Publish(bi.TransactionType(1), 10, 10)
+}
+
+func TestSubcribeOnceAsync(t *testing.T) {
+	results := make([]int, 0)
+
+	bus := New()
+	bus.SubscribeOnceAsync(bi.TransactionType(1), func(a int, out *[]int) {
+		*out = append(*out, a)
+	})
+
+	bus.Publish(bi.TransactionType(1), 10, &results)
+	bus.Publish(bi.TransactionType(1), 10, &results)
+
+	bus.WaitAsync()
+
+	if len(results) != 1 {
+		t.Fail()
+	}
+
+	if bus.HasCallback(bi.TransactionType(1)) {
+		t.Fail()
+	}
+}
+
+func TestSubscribeAsyncTransactional(t *testing.T) {
+	results := make([]int, 0)
+
+	bus := New()
+	bus.SubscribeAsync(bi.TransactionType(1), func(a int, out *[]int, dur string) {
+		sleep, _ := time.ParseDuration(dur)
+		time.Sleep(sleep)
+		*out = append(*out, a)
+	}, true)
+
+	bus.Publish(bi.TransactionType(1), 1, &results, "1s")
+	bus.Publish(bi.TransactionType(1), 2, &results, "0s")
+
+	bus.WaitAsync()
+
+	if len(results) != 2 {
+		t.Fail()
+	}
+
+	if results[0] != 1 || results[1] != 2 {
+		t.Fail()
+	}
+}
+
+func TestSubscribeAsync(t *testing.T) {
+	results := make(chan int)
+
+	bus := New()
+	bus.SubscribeAsync(bi.TransactionType(1), func(a int, out chan<- int) {
+		out <- a
+	}, false)
+
+	bus.Publish(bi.TransactionType(1), 1, results)
+	bus.Publish(bi.TransactionType(1), 2, results)
+
+	var numResults int32
+
+	go func() {
+		for _ = range results {
```

Comment:
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
should omit values from range; this loop is equivalent to `for range ...`